### PR TITLE
get_cert_by_subject : backward compatibility

### LINKED ISSUE #13565: Loading certificates on Android using unexpected hash
I've built an app on Android which bundles curl with openssl. This version of the app was able to execute HTTP requests, but not HTTPS, due to certificate authentication. After some investigation, I reached https://stackoverflow.com/questions/26935662. Which states:

> Android cacerts are generated with an md5 hash, but as of version 1.0.0, openssl uses sha1.

Note that the issue is not the location of the certificates (which you can define when building curl, I used `/system/etc/security/cacerts`).

Running `adb shell ls /system/etc/security/cacerts` on my device yields (I mean to say the certificates are there):
```
00673b5b.0
04f60c28.0
... 
```
I am presuming that the filenames or the file contents have something that make openssl unhappy (I depend on https://github.com/openssl/openssl/commit/719bc0e826317d22c0687cf919e97749fa53d54a ). In case this gives you any info, a certificate on Android looks like so:
```
::::::::::::::
/system/etc/security/cacerts/fde84897.0
::::::::::::::
-----BEGIN CERTIFICATE-----
...
-----END CERTIFICATE-----
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number:
            ...
        Signature Algorithm: sha1WithRSAEncryption
        Issuer: C=FR, O=Dhimyotis, CN=Certigna
        Validity
            Not Before: Jun 29 15:13:05 2007 GMT
            Not After : Jun 29 15:13:05 2027 GMT
        Subject: C=FR, O=Dhimyotis, CN=Certigna
        Subject Public Key Info:
            Public Key Algorithm: rsaEncryption
                RSA Public-Key: (2048 bit)
                Modulus:
                    ...
                Exponent: 65537 (0x10001)
        X509v3 extensions:
            X509v3 Basic Constraints: critical
                CA:TRUE
            X509v3 Subject Key Identifier: 
                ...
            X509v3 Authority Key Identifier: 
                keyid:...
                DirName:/C=FR/O=Dhimyotis/CN=Certigna
                serial:...

            X509v3 Key Usage: critical
                Certificate Sign, CRL Sign
            Netscape Cert Type: 
                SSL CA, S/MIME CA, Object Signing CA
    Signature Algorithm: sha1WithRSAEncryption
         ...
SHA1 Fingerprint=...
```
As mentioned above, most material available states that the issue is a mismatch between the hash algorithm used in the filenames, but nobody mentions a solution not involving side-loading a `cacert.pem` (like https://stackoverflow.com/questions/25253823), which isn't something I'm willing to do.

The solution I'm currently going with is using boringssl (instead of openssl, but still with curl) which works on Android out of the box, but I can't avoid feeling that I missed a flag like 

```
./Configure --use-md5-cert-filename-hash
```

when building openssl, or similar. Something which configures openssl to allow Android (or just old certificates in general? not sure it's specific to Android) certificates. Any input on this?

### PR DESCRIPTION
**Pull request by brunotl**
Fallback to MD5 name hash if no certificate found using SHA1 hash

fixes #13565 #15154

### CODE REVIEW

**Comment by t8m** [comment]
Not sure this is the right approach. But I cannot think of anything better. On the other hand the MD5 hash format was replaced with SHA1 quite long time ago. Do we really need to support it?

The CI is relevant, also we would need CLA signed to be able to take this contribution.

**Comment by brunotl** [comment]
> Not sure this is the right approach. But I cannot think of anything better. On the other hand the MD5 hash format was replaced with SHA1 quite long time ago. Do we really need to support it?

android  CA store use MD5, so it's  need to allow to use curl + openssl on android ...

> 
> The CI is relevant, also we would need CLA signed to be able to take this contribution.

i will look for CI and CLA...

**Comment by t8m** [comment]
@openssl/otc Any opinions on this?

**Comment by sizeak** [comment]
I've just run into this problem trying to use curl with https on Android, support for this would be greatly appreciated!

**Comment by t8m** [comment]
**OTC does not object to this feature.**

File: crypto/x509/by_dir.c [hunk 1]
```diff
@@ -257,165 +258,176 @@ static int get_cert_by_subject_ex(X509_LOOKUP *xl, X509_LOOKUP_TYPE type,
...
-        j = strlen(ent->dir) + 1 + 8 + 6 + 1 + 1;
-        if (!BUF_MEM_grow(b, j)) {
-            ERR_raise(ERR_LIB_X509, ERR_R_BUF_LIB);
-            goto finish;
+
+    hash[0] = X509_NAME_hash_ex(name, libctx, propq, &l); /* SHA1 name hash */
+    if (l == 0) {
+        hash[0] = 0;
+    }
+    hash[1] = X509_NAME_hash_old(name); /* MD5 name hash */
```
Comment (t-j-h) [inline crypto/x509/by_dir.c:266]: #ifndef OPENSSL_NO_MD5
required around here - as there are builds without MD5 included

Comment (brunotl) [reply crypto/x509/by_dir.c:266]: ok, i do that

**Comment by t8m** [comment]
Would it be possible to add a testcase?

**Comment by brunotl** [comment]
> Would it be possible to add a testcase?

i don't know how to do it...

**Comment by brunotl** [comment]
> You would add a test case to the test/recipes/60-test_x509_store.t Perl recipe. You could copy the last existing test and just slightly modify it. Instead of rehashing the dir with the copied root cert, you would copy the root cert with a name that would be the old name hash.

test case added.

File: crypto/x509/by_dir.c [hunk 2]
```diff
@@ -257,165 +258,180 @@ static int get_cert_by_subject_ex(X509_LOOKUP *xl, X509_LOOKUP_TYPE type,
...
-        ent = sk_BY_DIR_ENTRY_value(ctx->dirs, i);
-        j = strlen(ent->dir) + 1 + 8 + 6 + 1 + 1;
-        if (!BUF_MEM_grow(b, j)) {
-            ERR_raise(ERR_LIB_X509, ERR_R_BUF_LIB);
-            goto finish;
+
+    hash[0] = X509_NAME_hash_ex(name, libctx, propq, &l); /* SHA1 name hash */
+    if (l == 0) {
+        hash[0] = 0;
+    }
```
Comment (t8m) [inline crypto/x509/by_dir.c:265]: You cannot use 0 hash value as indicator for hash not present/invalid. Although the probability is just 1 in 4 billion, there can be a valid 0 hash value.

Please use a separate indicator value.

File: crypto/x509/by_dir.c [hunk 3]
```diff
@@ -257,165 +258,180 @@ static int get_cert_by_subject_ex(X509_LOOKUP *xl, X509_LOOKUP_TYPE type,
...
+    hash[1] = X509_NAME_hash_old(name); /* MD5 name hash */
+#else
+    hash[1] = 0;
+#endif
+
+    for (l = 0; l < 2; ++l) {
+        h = hash[l];
+        if (h == 0) {
+            continue; /* skip invalid hash */
         }
```
Comment (t8m) [inline crypto/x509/by_dir.c:276]: Please do not add `{}` around a single statement.

File: crypto/x509/by_dir.c [hunk 4]
```diff
@@ -257,165 +258,180 @@ static int get_cert_by_subject_ex(X509_LOOKUP *xl, X509_LOOKUP_TYPE type,
...
+
+    #ifdef OPENSSL_SYS_VMS
+                c = ent->dir[strlen(ent->dir) - 1];
+                if (c != ':' && c != '>' && c != ']') {
+                    /*
+                    * If no separator is present, we assume the directory
+                    * specifier is a logical name, and add a colon.  We really
+                    * should use better VMS routines for merging things like
+                    * this, but this will do for now... -- Richard Levitte
+                    */
```
Comment (t8m) [inline crypto/x509/by_dir.c:317]: Nit: indentation is off by one character here.

File: crypto/x509/by_dir.c [hunk 5]
```diff
@@ -257,165 +258,180 @@ static int get_cert_by_subject_ex(X509_LOOKUP *xl, X509_LOOKUP_TYPE type,
...
+                    BIO_snprintf(b->data, b->max,
+                                "%s%c%08lx.%s%d", ent->dir, c, h, postfix, k);
+                }
+    #ifndef OPENSSL_NO_POSIX_IO
+    # ifdef _WIN32
+    #  define stat _stat
+    # endif
+                {
+                    struct stat st;
+                    if (stat(b->data, &st) < 0)
```
Comment (t8m) [inline crypto/x509/by_dir.c:342]: As you're touching this (although it was preexisting) - please add one empty line after declarations.

File: crypto/x509/by_dir.c [hunk 6]
```diff
@@ -257,165 +258,180 @@ static int get_cert_by_subject_ex(X509_LOOKUP *xl, X509_LOOKUP_TYPE type,
...
-             */
-            if (hent == NULL) {
-                htmp.hash = h;
-                idx = sk_BY_DIR_HASH_find(ent->hashes, &htmp);
-                hent = sk_BY_DIR_HASH_value(ent->hashes, idx);
+            * we have added it to the cache so now pull it out again
+            *
+            * Note: quadratic time find here since the objects won't generally be
+            *       sorted and sorting the would result in O(n^2 log n) complexity.
+            */
```
Comment (t8m) [inline crypto/x509/by_dir.c:364]: Add one more space to indentation here.

File: crypto/x509/by_dir.c [hunk 7]
```diff
@@ -257,165 +258,180 @@ static int get_cert_by_subject_ex(X509_LOOKUP *xl, X509_LOOKUP_TYPE type,
...
-                hent = OPENSSL_malloc(sizeof(*hent));
-                if (hent == NULL) {
-                    CRYPTO_THREAD_unlock(ctx->lock);
-                    ok = 0;
+            /*
+            * If a CRL, update the last file suffix added for this.
+            * We don't need to add an entry if k is 0 as this is the initial value.
+            * This avoids the need for a write lock and sort operation in the
+            * simple case where no CRL is present for a hash.
+            */
```
Comment (t8m) [inline crypto/x509/by_dir.c:379]: Indentation as above.

File: crypto/x509/by_dir.c [hunk 8]
```diff
@@ -257,165 +258,180 @@ static int get_cert_by_subject_ex(X509_LOOKUP *xl, X509_LOOKUP_TYPE type,
...
+            if (type == X509_LU_CRL && k > 0) {
+                if (!CRYPTO_THREAD_write_lock(ctx->lock))
                     goto finish;
+                /*
+                * Look for entry again in case another thread added an entry
+                * first.
+                */
```
Comment (t8m) [inline crypto/x509/by_dir.c:386]: Nit: indentation

File: crypto/x509/by_dir.c [hunk 9]
```diff
@@ -257,165 +258,180 @@ static int get_cert_by_subject_ex(X509_LOOKUP *xl, X509_LOOKUP_TYPE type,
...
+                        OPENSSL_free(hent);
+                        ERR_raise(ERR_LIB_X509, ERR_R_CRYPTO_LIB);
+                        ok = 0;
+                        goto finish;
+                    }
+
+                    /*
+                    * Ensure stack is sorted so that subsequent sk_BY_DIR_HASH_find
+                    * will not mutate the stack and therefore require a write lock.
+                    */
```
Comment (t8m) [inline crypto/x509/by_dir.c:412]: Nit: indentation

File: crypto/x509/by_dir.c [hunk 10]
```diff
@@ -257,165 +258,180 @@ static int get_cert_by_subject_ex(X509_LOOKUP *xl, X509_LOOKUP_TYPE type,
...
 
-            /*
-             * Clear any errors that might have been raised processing empty
-             * or malformed files.
-             */
-            ERR_clear_error();
+                /*
+                * Clear any errors that might have been raised processing empty
+                * or malformed files.
+                */
```
Comment (t8m) [inline crypto/x509/by_dir.c:430]: Nit: indentation

File: test/recipes/60-test_x509_store_old.t [hunk 11]
```diff
@@ -0,0 +1,53 @@
+#! /usr/bin/env perl
+# Copyright 2015-2016 The OpenSSL Project Authors. All Rights Reserved.
```
Comment (t8m) [inline test/recipes/60-test_x509_store_old.t:2]: Just 2024

File: crypto/x509/by_dir.c [hunk 12]
```diff
@@ -257,165 +258,180 @@ static int get_cert_by_subject_ex(X509_LOOKUP *xl, X509_LOOKUP_TYPE type,
...
-                                            propq)) == 0)
-                    break;
-            } else if (type == X509_LU_CRL) {
-                if ((X509_load_crl_file(xl, b->data, ent->dir_type)) == 0)
-                    break;
+                if (c == '\0') {
+                    /*
+                    * This is special.  When c == '\0', no directory separator
+                    * should be added.
+                    */
```
Comment (t8m) [inline crypto/x509/by_dir.c:327]: Nit: indentation

**Review by t8m — CHANGES_REQUESTED** [review]

File: crypto/x509/by_dir.c (hunk 5 above)
Comment (brunotl) [reply crypto/x509/by_dir.c:342]: ef74616d2ea2cafcfe6c9cb04e00b69b0f6e8c6c

File: crypto/x509/by_dir.c (hunk 4 above)
Comment (brunotl) [reply crypto/x509/by_dir.c]: ef74616d2ea2cafcfe6c9cb04e00b69b0f6e8c6c

File: crypto/x509/by_dir.c (hunk 3 above)
Comment (brunotl) [reply crypto/x509/by_dir.c]: 4612aaf293430102137904d73096f496e642fe43

File: crypto/x509/by_dir.c (hunk 8 above)
Comment (brunotl) [reply crypto/x509/by_dir.c]: ef74616d2ea2cafcfe6c9cb04e00b69b0f6e8c6c

File: crypto/x509/by_dir.c (hunk 10 above)
Comment (brunotl) [reply crypto/x509/by_dir.c]: ef74616d2ea2cafcfe6c9cb04e00b69b0f6e8c6c

File: test/recipes/60-test_x509_store_old.t (hunk 11 above)
Comment (brunotl) [reply test/recipes/60-test_x509_store_old.t]: 1301c593f88d5ea4d7087e72b42bd78bbfc2e344

**Review by t8m — APPROVED** [review]

File: crypto/x509/by_dir.c [hunk 13]
```diff
@@ -257,165 +259,179 @@ static int get_cert_by_subject_ex(X509_LOOKUP *xl, X509_LOOKUP_TYPE type,
...
+    hash[0] = X509_NAME_hash_ex(name, libctx, propq, &l); /* SHA1 name hash */
+    valid_hash[0] = (l != 0);
+#ifndef OPENSSL_NO_MD5    
+    hash[1] = X509_NAME_hash_old(name); /* MD5 name hash */
+    valid_hash[1] = 1;
+#else
+    valid_hash[1] = 0;
+#endif
+
+    for (l = 0; l < 2; ++l) {
```
[the remark this answers is not in the document]
Comment (brunotl) [reply crypto/x509/by_dir.c:272]: You are right for sure, but that requires too much refactoring and too much regression risk for me. If you want to do this refactoring, you are welcome to proceed.

**Comment by t8m** [comment]
ping for second review

File: crypto/x509/by_dir.c (hunk 13 above)
[the remark this answers is not in the document]
Comment (FdaSilvaYY) [reply crypto/x509/by_dir.c:272]: @brunotl do you mind if I push a branch with a refactoring and your changes ?

[the remark this answers is not in the document]
Comment (brunotl) [reply crypto/x509/by_dir.c:272]: @FdaSilvaYY  no problem, you can !

**Comment by vdukhovni** [comment]
I'd like to not rush to merge this PR.  The *CApath* (`by_dir`) lookup method looks semantically questionable to me as a consequence of the internal cache. This should IMNSHO always read the directory anew, so that changes made by `rehash` are seen even by long running servers.

But for that we'd also want to make `rehash` more "atomic", as failure to update the directory content in a manner that ensures that certs that are present before and after the rehash are always visible *during* the rehash would be more likely to happen once the cache is removed.

Support for MD5 hashes could be added once the larger issues are addressed, but in the meantime, I'd rather not see the code get more complex.

I'm admittedly assuming that the idea of dropping the cache and improving robustness during rehash is generally supported.

@levitte in particular may be able to sanity check the above rant.

**Comment by sizeak** [comment]
Hi there,

I currently maintain an internal fork of this code (well, something similar / simpler) to support Android builds. I appreciate that avoiding some extra code complexity now could make future refactoring easier, but as a consumer of the library I would really appreciate it if this PR or something similar was merged soon so that I no longer needed to maintain a fork.

While I am just an external observer to this discussion, the changes you've suggested sound sensible, but they also seem like a future iteration on the code, and blocking this PR until they are implemented would be a shame because this PR has immediate, real world benefits to multiple consumers. i.e. it really would make my life easier. Certainly they shouldn't be a part of _this_ PR.

OTOH I assume the `by_dir` situation isn't a new thing and could wait, maybe it's even YAGNI thing. 🤷🏻 Either way, as I understand it, merging this won't make the functional situation worse in that regard, but as mentioned will make other things better.

**Comment by vdukhovni** [comment]
Support for the legacy lookup was dropped many years ago, in a release that is now EOL. I am rather sceptical there's much demand for the legacy lookup. Anyone who's going to deploy a new enough OpenSSL build that incorporates these changes could more easily just use an extant built to generate the supported hashes instead.

So overall, I don't think this use case warrants the new code.

**Comment by t8m** [comment]
> Support for the legacy lookup was dropped many years ago, in a release that is now EOL. I am rather sceptical there's much demand for the legacy lookup. Anyone who's going to deploy a new enough OpenSSL build that incorporates these changes could more easily just use an extant built to generate the supported hashes instead.
> 
> So overall, I don't think this use case warrants the new code.

@vdukhovni There is one big use-case for the legacy lookup. And that is namely for the Android installs that want to do lookups in the system certificate store. That is using the legacy MD-5 hash names.

One option would be to make this code if-defed out and built only on Android. Or even simplify this patch to just switch between the hash used based on a macro such as OPENSSL_LEGACY_CERT_HASH.

**Comment by vdukhovni** [comment]
I have no objections to (indeed support) a compile-time configurable **single** hash algorithm for the `by_dir` (*CApath*) trust store. If there are no objections, let's do that instead.

**Comment by petrovr** [comment]
На 15.10.24 г. в 10:06 ч., Tomáš Mráz написа:
>> Support for the legacy lookup was dropped many years ago, in a release that is now EOL. I am rather sceptical there's much demand for the legacy lookup. Anyone who's going to deploy a new enough OpenSSL build that incorporates these changes could more easily just use an extant built to generate the supported hashes instead.
>>
>> So overall, I don't think this use case warrants the new code.
> 
> @vdukhovni There is one big use-case for the legacy lookup. And that is namely for the Android installs that want to do lookups in the system certificate store. That is using the legacy MD-5 hash names.
> 
> One option would be to make this code if-defed out and built only on Android. Or even simplify this patch to just switch between the hash used based on a macro such as OPENSSL_LEGACY_CERT_HASH.


Android settings allows user to enable or disable root certificates. 
Lookup works fine if is from application(java) code i.e. from 
java.security.KeyStore.getInstance("AndroidCAStore").

To ensure proper functionality a specific OpenSSL X509 lookup should 
redirect request to java key-store. This is correct implementation.

It is possible i.e., SecureBox Pro 3.0.1 adds following functionality 
"Look-up for trusted(root) certificates installed on device - used in 
verification of certificate based identities or host-keys.".
PKIX-SSH uses OpenSSL engine to access java key-store.


It is possible to use "by dir" lookup (with specific hash algorithm) but 
this will find CA-certificates disabled by device settings.


Roumen

**Comment by vdukhovni** [comment]
> It is possible to use "by dir" lookup (with specific hash algorithm) but this will find CA-certificates disabled by device settings. Roumen

I believe you're saying that direct use of `by_dir` is not (no longer?) a good idea on Android, and that well integrated applications should somehow find a way to use the system JKS as a source of trust anchors (at least by default?).

If so, this leads to some followup questions:

- Do some applications specifically ship and use a custom set of trust anchors, that meets their needs, that are deliberately separate from the system trust store?  [ If so, presumably they don't really need `md5` legacy hash support. ]
- Is there a well-maintained high quality open source "bridge" that would allow OpenSSL applications currently using the legacy trust store, to use the system JKS instead?
- If not, is there anyone in the OpenSSL on Android developer community willing to contribute such a "bridge" for inclusion in OpenSSL?

Bottom line, what are the best options, assuming there's enough community interest in OPenSSL on Android?

**Comment by petrovr** [comment]
For instance on Android 15 (API Level 35)  emulator from "Settings"->"Security & privacy"->"More security & privacy"->"Encryption & credentials"->"Trusted Certificates" user could exclude "system" CA certificate.

User could upload CA certificate. For instance image android-user-ca shows CA certificate used in PKIX-SSH regression test. Alias is highlighted, subject is in lower case.
Remark: Actual distinguished name is "C=XX,ST=World,L=Somewhere cyrillic-АБВ-Яабв-я greek-ΑΒΓ-Ωαβγ-ω,O=SSH Test Team cyrillic-АБВ-Яабв-я greek-ΑΒΓ-Ωαβγ-ω,OU=SSH Testers cyrillic-АБВ-Яабв-я greek-ΑΒΓ-Ωαβγ-ω,OU=SSH Testers cyrillic-АБВ-Яабв-я greek-ΑΒΓ-Ωαβγ-ω level 0,CN=SSH TestCA level 0"


So loop(enumerate) on jdk ca-store entries are with alias {system|user}:<HASH>.<N> where <N> is zero and <HASH> is  MD5 X509_NAME hash. 


To implement by-subject X.509 lookup Android application still need MD5 hash.

**Comment by petrovr** [comment]
Personally I'm not convinced that proposed change is valuable.

As exist function X509_NAME_hash_old i.e., could be implemented by-store based lookup.

----

Off-topic: I' do not have enough spare time to implement public JNI based lookup for Android.

**Comment by brunotl** [comment]
my goal behind this PR is only to allow to use libcurl with openssl for https without disabling server certificate verification on android ...

**Comment by sizeak** [comment]
> to allow to use libcurl with openssl for https
This is all I currently need it for too.

**Review by InfoHunter — COMMENTED** [review]
Since the 'to-be-resolved' original issues (#13565 and #15154) are closed as unresolved already. Do we consider to close this PR as well?

**Comment by t8m** [comment]
Yes, I am afraid there is not enough convincing arguments to take this further.
